# Import Packages

In [66]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from scipy.stats import zscore

%matplotlib inline



In [67]:
# Import Data
data = pd.read_csv('adult.csv')

# Quality Check

In [68]:
data.shape

(32560, 15)

In [69]:
data.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [70]:
#Name of the columns
data.columns = ['Age', 'Workclass', 'fnlgt', 'Education', 'Education_num', 'Marital_Status', 'Occupation'
               ,'Relationship', 'Race', 'Sex', 'Gain', 'Loss', 'Hoursperweek', 'Country', 'Label']

In [71]:
data.head()

Age          Workclass   fnlgt   Education  Education_num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   
3   28            Private  338409   Bachelors             13   
4   37            Private  284582     Masters             14   

        Marital_Status          Occupation    Relationship    Race      Sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   Gain  Loss  Hoursperweek         Country   Label  
0     0     0            13   United-States   <=50K  
1     0     0            40   United-States   <=50K  
2     0     0            40   United-States   <=50K  
3     0     0            40            Cuba   <=50K  
4     0     0            40   United-States   <=50K

In [72]:
# Data Types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32560 entries, 0 to 32559
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Age             32560 non-null  int64 
 1   Workclass       32560 non-null  object
 2   fnlgt           32560 non-null  int64 
 3   Education       32560 non-null  object
 4   Education_num   32560 non-null  int64 
 5   Marital_Status  32560 non-null  object
 6   Occupation      32560 non-null  object
 7   Relationship    32560 non-null  object
 8   Race            32560 non-null  object
 9   Sex             32560 non-null  object
 10  Gain            32560 non-null  int64 
 11  Loss            32560 non-null  int64 
 12  Hoursperweek    32560 non-null  int64 
 13  Country         32560 non-null  object
 14  Label           32560 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


Great! We don't have null values. Yes, my first thought! But.....

# Change label to 0 and 1
le = LabelEncoder()
data['Label'] = le.fit_transform(data['Label'])

In [74]:
le = LabelEncoder()
data['Label'] = le.fit_transform(data['Label'])

In [17]:
data.Label.value_counts()#/data.shape[0]

0    24719
1     7841
Name: Label, dtype: int64

# Preprocessing

In [75]:
# Normalize the numerical data

num = [var for var in data.columns if (data[var].dtype == 'int64') & (var != 'Label')]
cat = [var for var in data.columns if data[var].dtype == 'object']

In [76]:
num

['Age', 'fnlgt', 'Education_num', 'Gain', 'Loss', 'Hoursperweek']

In [77]:
normalized_data = zscore(data[num], axis=1)

In [78]:
normalized_data.head()

Age     fnlgt  Education_num      Gain      Loss  Hoursperweek
0 -0.446092  2.236068      -0.447284 -0.447703 -0.447703     -0.447284
1 -0.446957  2.236068      -0.447318 -0.447430 -0.447430     -0.446932
2 -0.446836  2.236068      -0.447362 -0.447442 -0.447442     -0.446985
3 -0.447120  2.236068      -0.447239 -0.447342 -0.447342     -0.447025
4 -0.447036  2.236068      -0.447253 -0.447385 -0.447385     -0.447008

In [79]:
## QC
np.mean(normalized_data), np.std(normalized_data)

(Age             -0.450492
 fnlgt            2.232898
 Education_num   -0.451105
 Gain            -0.431391
 Loss            -0.449455
 Hoursperweek    -0.450455
 dtype: float64,
 Age              0.019186
 fnlgt            0.046391
 Education_num    0.019260
 Gain             0.113947
 Loss             0.021548
 Hoursperweek     0.019194
 dtype: float64)

In [80]:
data_v1 = pd.concat([normalized_data, data[cat], data['Label']], axis = 1)

In [81]:
data_v1.head()

Age     fnlgt  Education_num      Gain      Loss  Hoursperweek  \
0 -0.446092  2.236068      -0.447284 -0.447703 -0.447703     -0.447284   
1 -0.446957  2.236068      -0.447318 -0.447430 -0.447430     -0.446932   
2 -0.446836  2.236068      -0.447362 -0.447442 -0.447442     -0.446985   
3 -0.447120  2.236068      -0.447239 -0.447342 -0.447342     -0.447025   
4 -0.447036  2.236068      -0.447253 -0.447385 -0.447385     -0.447008   

           Workclass   Education       Marital_Status          Occupation  \
0   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
1            Private     HS-grad             Divorced   Handlers-cleaners   
2            Private        11th   Married-civ-spouse   Handlers-cleaners   
3            Private   Bachelors   Married-civ-spouse      Prof-specialty   
4            Private     Masters   Married-civ-spouse     Exec-managerial   

     Relationship    Race      Sex         Country  Label  
0         Husband   White     Male   United-States      0  
1   Not-in-family   White     Male   United-States      0  
2         Husband   Black     Male   United-States      0  
3            Wife   Black   Female            Cuba      0  
4            Wife   White   Female   United-States      0

In [82]:
data_v1.shape

(32560, 15)

In [83]:
# Categorical varibales
cat

['Workclass',
 'Education',
 'Marital_Status',
 'Occupation',
 'Relationship',
 'Race',
 'Sex',
 'Country']

In [84]:
data_v1.Workclass.value_counts()

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1297
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: Workclass, dtype: int64

.....But, we do have Null values in our data, just not in a standard format. Let's deal with them!
Let's check for other categorical columns as well and replace these values with NaN

In [85]:
data_v1['Workclass'].replace('?', np.NaN, inplace=True)

In [86]:
data_v1.Workclass.value_counts()

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1297
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: Workclass, dtype: int64

Strange! It seems like they are not strings, so lets convert them to strings

In [87]:
data_v1['Workclass'] = data_v1['Workclass'].astype('string')

In [88]:
data_v1['Workclass'].replace('?', np.NaN, inplace=True)

In [89]:
data_v1.Workclass.value_counts()

Private             22696
Self-emp-not-inc     2541
Local-gov            2093
?                    1836
State-gov            1297
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: Workclass, dtype: Int64

Not working, Now there's only one thing left to do. remove these rows

In [90]:
rows = data_v1['Workclass'].str.contains(r"\?")

data_v1= data_v1.drop(rows[rows==True].index.to_list(), axis=0)

data_v1.shape

(30724, 15)

In [91]:
#for var in cat:
   # print(data_v1[var].value_counts())

In [92]:
rows = data_v1['Occupation'].str.contains(r"\?")

data_v1= data_v1.drop(rows[rows==True].index.to_list(), axis=0)

data_v1.shape

(30717, 15)

In [93]:
rows = data_v1['Country'].str.contains(r"\?")

data_v1= data_v1.drop(rows[rows==True].index.to_list(), axis=0)

data_v1.shape

(30161, 15)

In [94]:
data_v1.isna().sum()

Age               0
fnlgt             0
Education_num     0
Gain              0
Loss              0
Hoursperweek      0
Workclass         0
Education         0
Marital_Status    0
Occupation        0
Relationship      0
Race              0
Sex               0
Country           0
Label             0
dtype: int64

### Check for Cardinality

In [95]:
for var in cat:
    print(var+' contains '+str(data_v1[var].nunique())+' labels')
    print()

Workclass contains 7 labels

Education contains 16 labels

Marital_Status contains 7 labels

Occupation contains 14 labels

Relationship contains 6 labels

Race contains 5 labels

Sex contains 2 labels

Country contains 41 labels



In [96]:
data_v1['Country'].value_counts()

 United-States                 27503
 Mexico                          610
 Philippines                     188
 Germany                         128
 Puerto-Rico                     109
 Canada                          107
 India                           100
 El-Salvador                     100
 Cuba                             92
 England                          86
 Jamaica                          80
 South                            71
 China                            68
 Italy                            68
 Dominican-Republic               67
 Vietnam                          64
 Guatemala                        63
 Japan                            59
 Poland                           56
 Columbia                         56
 Iran                             42
 Taiwan                           42
 Haiti                            42
 Portugal                         34
 Nicaragua                        33
 Peru                             30
 Greece                           29
 

91% of the values are US, so we'll club all other countries in Other

In [97]:
list_of_countries = [Country for Country in data_v1.Country.unique() if Country != ' United-States']

In [98]:
data_v1['Country'] = data_v1['Country'].apply(lambda x: 'Other' if x in list_of_countries else x)

In [99]:
data_v1.Country.value_counts()

 United-States    27503
Other              2658
Name: Country, dtype: int64

In [100]:
cat

['Workclass',
 'Education',
 'Marital_Status',
 'Occupation',
 'Relationship',
 'Race',
 'Sex',
 'Country']

In [101]:
# One Hot Encoding

data_final = pd.get_dummies(data_v1, columns = cat)

In [102]:
data_final.head()

Age     fnlgt  Education_num      Gain      Loss  Hoursperweek  Label  \
0 -0.446092  2.236068      -0.447284 -0.447703 -0.447703     -0.447284      0   
1 -0.446957  2.236068      -0.447318 -0.447430 -0.447430     -0.446932      0   
2 -0.446836  2.236068      -0.447362 -0.447442 -0.447442     -0.446985      0   
3 -0.447120  2.236068      -0.447239 -0.447342 -0.447342     -0.447025      0   
4 -0.447036  2.236068      -0.447253 -0.447385 -0.447385     -0.447008      0   

   Workclass_ Federal-gov  Workclass_ Local-gov  Workclass_ Private  ...  \
0                       0                     0                   0  ...   
1                       0                     0                   1  ...   
2                       0                     0                   1  ...   
3                       0                     0                   1  ...   
4                       0                     0                   1  ...   

   Relationship_ Wife  Race_ Amer-Indian-Eskimo  Race_ Asian-Pac-Islander  \
0                   0                         0                         0   
1                   0                         0                         0   
2                   0                         0                         0   
3                   1                         0                         0   
4                   1                         0                         0   

   Race_ Black  Race_ Other  Race_ White  Sex_ Female  Sex_ Male  \
0            0            0            1            0          1   
1            0            0            1            0          1   
2            1            0            0            0          1   
3            1            0            0            1          0   
4            0            0            1            1          0   

   Country_ United-States  Country_Other  
0                       1              0  
1                       1              0  
2                       1              0  
3                       0              1  
4                       1              0  

[5 rows x 66 columns]

In [103]:
data_final.shape

(30161, 66)

In [104]:
data_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30161 entries, 0 to 32559
Data columns (total 66 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Age                                    30161 non-null  float64
 1   fnlgt                                  30161 non-null  float64
 2   Education_num                          30161 non-null  float64
 3   Gain                                   30161 non-null  float64
 4   Loss                                   30161 non-null  float64
 5   Hoursperweek                           30161 non-null  float64
 6   Label                                  30161 non-null  int64  
 7   Workclass_ Federal-gov                 30161 non-null  uint8  
 8   Workclass_ Local-gov                   30161 non-null  uint8  
 9   Workclass_ Private                     30161 non-null  uint8  
 10  Workclass_ Self-emp-inc                30161 non-null  uint8  
 11  Wo

# Splitting the Data

In [105]:
X = data_final.drop(columns = 'Label')
Y = data_final['Label']

In [106]:
X.shape, Y.shape

((30161, 65), (30161,))

In [107]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.1, random_state = 42)

In [108]:
X_train.shape, X_test.shape

((27144, 65), (3017, 65))

In [109]:
X_train.head()

Age     fnlgt  Education_num      Gain      Loss  Hoursperweek  \
31438 -0.447169  2.236068      -0.447241 -0.447291 -0.447291     -0.447076   
7280  -0.444323  2.236064      -0.448223 -0.449021 -0.449021     -0.445475   
31021 -0.446848  2.236068      -0.447341 -0.447442 -0.447442     -0.446994   
30953 -0.484874  2.230218      -0.486059 -0.287675 -0.486334     -0.485276   
24047 -0.446908  2.236068      -0.447284 -0.447444 -0.447444     -0.446988   

       Workclass_ Federal-gov  Workclass_ Local-gov  Workclass_ Private  \
31438                       0                     0                   1   
7280                        0                     0                   1   
31021                       0                     0                   1   
30953                       0                     0                   1   
24047                       0                     0                   0   

       Workclass_ Self-emp-inc  ...  Relationship_ Wife  \
31438                        0  ...                   0   
7280                         0  ...                   0   
31021                        0  ...                   0   
30953                        0  ...                   0   
24047                        0  ...                   0   

       Race_ Amer-Indian-Eskimo  Race_ Asian-Pac-Islander  Race_ Black  \
31438                         0                         0            0   
7280                          0                         0            0   
31021                         0                         0            0   
30953                         0                         0            1   
24047                         0                         0            0   

       Race_ Other  Race_ White  Sex_ Female  Sex_ Male  \
31438            0            1            0          1   
7280             0            1            1          0   
31021            0            1            1          0   
30953            0            0            0          1   
24047            0            1            0          1   

       Country_ United-States  Country_Other  
31438                       1              0  
7280                        1              0  
31021                       1              0  
30953                       1              0  
24047                       1              0  

[5 rows x 65 columns]

# Model Training

In [110]:
from sklearn.naive_bayes import GaussianNB

# Instantiate the model
gnb = GaussianNB()


# Fit to train data
gnb.fit(X_train, Y_train)

GaussianNB()

In [111]:
from sklearn.metrics import accuracy_score

acc_train = accuracy_score(Y_train, gnb.predict(X_train))
acc_test = accuracy_score(Y_test, gnb.predict(X_test))

In [112]:
acc_train, acc_test

(0.6971706454465075, 0.6920782234007292)

The training-set accuracy score is 0.697 while the test-set accuracy to be 0.711. These two values are quite comparable. So, there is no sign of overfitting.